File: app/utils/sync.py
"""
Cloud sync for MyDailyOps offline-first system.

Handles bidirectional sync between SQLite cache and Supabase.
"""

import sqlite3
import json
from app.supabase.client import supabase
from app.database.offline import (
    cache_replace_all,
    pop_pending_updates,
    get_pending_count,
    DB_FILE
)
from kivy.app import App


def pull_from_supabase(user_id):
    """
    Pull tasks from Supabase and update local cache.
    
    Args:
        user_id: Current user ID
        
    Returns:
        True if successful, False otherwise
    """
    try:
        print("⬇️ Pulling tasks from Supabase...")
        
        # Fetch all tasks from Supabase
        response = supabase.table("tasks") \
            .select("*") \
            .eq("user_id", user_id) \
            .eq("deleted", False) \
            .order("created_at", desc=True) \
            .execute()
        
        if response.data:
            tasks = response.data
            
            # Update local cache with server data
            cache_replace_all(tasks)
            
            print(f"✅ Pulled {len(tasks)} tasks from Supabase")
            return True
        else:
            print("ℹ️ No tasks found on server")
            return True
            
    except Exception as e:
        print(f"❌ Error pulling from Supabase: {e}")
        return False


def push_to_supabase():
    """
    Push pending updates to Supabase.
    
    Returns:
        True if successful, False otherwise
    """
    try:
        pending_count = get_pending_count()
        
        if pending_count == 0:
            print("ℹ️ No pending updates to push")
            return True
        
        print(f"⬆️ Pushing {pending_count} updates to Supabase...")
        
        # Get all pending updates
        updates = pop_pending_updates()
        
        success_count = 0
        failed_updates = []
        
        for update in updates:
            operation = update["operation"]
            task_id = update["task_id"]
            
            try:
                if operation == "create":
                    # Insert new task
                    payload = update["payload"]
                    response = supabase.table("tasks").insert(payload).execute()
                    if response.data:
                        success_count += 1
                        print(f"  ✅ Created: {payload.get('title')}")
                    
                elif operation == "update":
                    # Update existing task
                    payload = update["payload"]
                    response = supabase.table("tasks") \
                        .update(payload) \
                        .eq("id", task_id) \
                        .execute()
                    if response.data:
                        success_count += 1
                        print(f"  ✅ Updated: {task_id}")
                    
                elif operation == "delete":
                    # Soft delete task
                    response = supabase.table("tasks") \
                        .update({"deleted": True}) \
                        .eq("id", task_id) \
                        .execute()
                    if response.data:
                        success_count += 1
                        print(f"  ✅ Deleted: {task_id}")
                        
            except Exception as e:
                print(f"  ❌ Failed {operation} for {task_id}: {e}")
                failed_updates.append(update)
        
        # Re-queue failed updates
        if failed_updates:
            conn = sqlite3.connect(DB_FILE)
            cursor = conn.cursor()
            for update in failed_updates:
                cursor.execute("""
                    INSERT INTO pending_updates (task_id, operation, payload, timestamp)
                    VALUES (?, ?, ?, ?)
                """, (
                    update["task_id"],
                    update["operation"],
                    json.dumps(update["payload"]) if update["operation"] != "delete" else update["payload"],
                    update["timestamp"]
                ))
            conn.commit()
            conn.close()
            print(f"⚠️ Re-queued {len(failed_updates)} failed updates")
        
        print(f"✅ Push complete: {success_count}/{len(updates)} succeeded")
        return success_count > 0 or len(updates) == 0
        
    except Exception as e:
        print(f"❌ Error pushing to Supabase: {e}")
        return False


def sync_now(user_id):
    """
    Perform full bidirectional sync: push → pull.
    
    Args:
        user_id: Current user ID
        
    Returns:
        True if successful, False otherwise
    """
    print("\n🔄 Starting sync...")
    
    # Step 1: Push local changes to server
    push_success = push_to_supabase()
    
    # Step 2: Pull latest from server
    pull_success = pull_from_supabase(user_id)
    
    if push_success and pull_success:
        print("✅ Sync completed successfully\n")
        return True
    else:
        print("⚠️ Sync completed with errors\n")
        return False

File: mobile/lib/sync.ts
import { supabase, getCurrentUserId } from './supabase';
import {
  loadTasksFromCache,
  upsertTaskToCache,
  replaceAllTasksInCache,
} from '../database/init';
import { Task } from '../types/task';

/**
 * Pull all tasks from Supabase for current user and replace local cache
 */
export async function pullFromSupabase(userId: string): Promise<Task[]> {
  console.log('[Sync] Pulling tasks for user:', userId);

  try {
    const { data, error } = await supabase
      .from('tasks')
      .select('*')
      .eq('user_id', userId)
      .order('updated_at', { ascending: false });

    if (error) throw error;

    const tasks: Task[] = (data || []).map((row: any) => ({
      id: row.id,
      user_id: row.user_id,
      title: row.title,
      description: row.description || '',
      priority: row.priority,
      category: row.category,
      deadline: row.deadline,
      status: row.status,
      pinned: row.pinned || false,
      created_at: row.created_at,
      updated_at: row.updated_at,
    }));

    // Replace entire local cache with server data
    await replaceAllTasksInCache(userId, tasks);

    console.log(`[Sync] Pulled ${tasks.length} tasks from Supabase`);
    return tasks;
  } catch (error) {
    console.error('[Sync] Error pulling from Supabase:', error);
    throw error;
  }
}

/**
 * Push a single task to Supabase (upsert)
 */
export async function pushTaskToSupabase(task: Task): Promise<void> {
  console.log('[Sync] Pushing task to Supabase:', task.id);

  try {
    const { error } = await supabase
      .from('tasks')
      .upsert({
        id: task.id,
        user_id: task.user_id,
        title: task.title,
        description: task.description,
        priority: task.priority,
        category: task.category,
        deadline: task.deadline,
        status: task.status,
        pinned: task.pinned,
        created_at: task.created_at,
        updated_at: task.updated_at,
      });

    if (error) throw error;

    console.log('[Sync] Task pushed successfully:', task.id);
  } catch (error) {
    console.error('[Sync] Error pushing task to Supabase:', error);
    throw error;
  }
}

/**
 * Delete task from Supabase
 */
export async function deleteTaskFromSupabase(taskId: string, userId: string): Promise<void> {
  console.log('[Sync] Deleting task from Supabase:', taskId);

  try {
    const { error } = await supabase
      .from('tasks')
      .delete()
      .eq('id', taskId)
      .eq('user_id', userId);

    if (error) throw error;

    console.log('[Sync] Task deleted successfully:', taskId);
  } catch (error) {
    console.error('[Sync] Error deleting task from Supabase:', error);
    throw error;
  }
}

/**
 * Full sync: pull all tasks from Supabase
 * Returns tasks from cache if sync fails
 */
export async function syncNow(): Promise<Task[]> {
  console.log('[Sync] Starting full sync...');

  try {
    const userId = await getCurrentUserId();
    if (!userId) {
      console.warn('[Sync] Not authenticated, returning empty array');
      return [];
    }

    console.log('[Sync] Authenticated as user:', userId);

    // Pull all tasks from server (replaces local cache)
    const tasks = await pullFromSupabase(userId);

    console.log('[Sync] Full sync completed successfully, fetched', tasks.length, 'tasks');
    return tasks;
  } catch (error) {
    console.error('[Sync] Full sync failed:', error);
    
    // If sync fails, try to return local cache
    const userId = await getCurrentUserId();
    if (userId) {
      const cachedTasks = await loadTasksFromCache(userId);
      console.log('[Sync] Returning', cachedTasks.length, 'cached tasks');
      return cachedTasks;
    }
    
    return [];
  }
}
